### PR TITLE
Translate numericCast

### PR DESCRIPTION
**Pull request by leopic**
Happy new year folks 👋 

I have a new translation for you to review, this time it's `numericCast`'s turn.

Looking forward to your comments.

Thanks,
Leo

### CODE REVIEW

File: 2018-10-15-numericcast.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
+---
+title: numericCast(_:)
+author: Mattt
+translator: Leo Picado
+category: Swift
+excerpt: >
+  Lograr que el código compile es distinto a hacer las cosas bien, pero a veces ocupamos lo primero para llegar a lo segundo.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: a veces necesitamos?

File: 2018-10-15-numericcast.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+status:
+  swift: 4.2
+---
+
+Todos tenemos nuestra analogía preferida para describir la programación, desde carpintería, bordado o hasta jardinería. Quizás lo reducimos a resolver problemas, contar una historia o quizás solo como un arte.
+
+Programar es muy comparable a escribir, la pregunta es ¿es prosa o poesía? ¿Y si fuera música? ¿Qué género sería? Sería jazz, sin duda alguna.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: muy similar?

File: 2018-10-15-numericcast.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+Tal vez los cuentos del medio oeste brindan la comparación más cercana sobre lo que hacemos a diario. Al ojear cualquier edición de _Las mil y una noches_ (<bdi lang="ar">أَلْف لَيْلَة وَلَيْلَة</bdi>) vemos descripciones de entes sobrenaturales conocidos como <dfn>geniecillos</dfn>, <dfn>genios</dfn> o 🧞. Sin importar como se les llame todos estamos familiarizados con su habilidad de conceder deseos y las tragedias que inevitablemente causan.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: brinden la comparación..

Qué tal:
Sin importar cómo se les llame, todos estamos familiarizados con su habilidad de conceder deseos y con las tragedias que siempre causan.

File: 2018-10-15-numericcast.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+Podríamos decir que de muchas maneras, las computadoras representan una constitución física de un deseo. Una computadora, al igual que un genio, felizmente cumplirá cualquier deseo solicitado, sin miramiento alguno y para cuando nos demos cuenta que se cometió un error, puede ser demasiado tarde para solucionarlo.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: decir que, de muchas maneras, las computadoras...

File: 2018-10-15-numericcast.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+Como desarrolladores en Swift es muy probable que en algún momento hayamos caído presa de errores al intentar convertir tipos enteros y que también, hayamos pensados "como desearía que estos errores desaparecieran y mi código finalmente compilara".
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: , hayamos deseado que los errores desaparecieran y el código compilara.

File: 2018-10-15-numericcast.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+Si eso le suena familiar recibirá con agrado el tema de esta semana, se trata de `numericCast(_:)`, una pequeña función incluida en la librería estándar del lenguaje que puede ser justamente lo que tanto deseaba. Sin embargo siempre debemos ser cuidadosos con lo que deseamos, porque podría convertirse en realidad.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: recibirás? (tuteo)
lo que tanto deseabas

File: 2018-10-15-numericcast.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+Desmitifiquemos a `numericCast(_:)` [viendo su implementación](https://github.com/apple/swift/blob/7f7b4f12d3138c5c259547c49c3b41415cd4206e/stdlib/public/core/Integers.swift#L3508-L3510):
+
+```swift
+public func numericCast<T : BinaryInteger, U : BinaryInteger>(_ x: T) -> U {
+  return U(x)
+}
+```
+
+(Tal y como aprendimos en [nuestro artículo sobre Never](https://nshipster.es/never/), hasta la cantidad más pequeña de Swift puede tener un gran impacto).
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: la cosa más pequeña

File: 2018-10-15-numericcast.md [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+El protocolo [`BinaryInteger`](https://developer.apple.com/documentation/swift/binaryinteger) fue introducido en Swift 4 como parte de un esfuerzo para unificar la interfaz y el manejo de enteros, tanto con y sin signo, sin importar su talla y tamaño.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: su tamaño?

File: 2018-10-15-numericcast.md [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+  ([`init?(exactly:)`](https://developer.apple.com/documentation/swift/binaryinteger/2925955-init)):
+  Devuelve `nil` para valores fuera de rango
+- **Conversión de prensa**
+  ([`init(clamping:)`](https://developer.apple.com/documentation/swift/binaryinteger/2886143-init)):
+  Usa el valor más cercano para representar valores fuera de rango
+- **Conversión de patrón de bits**
+  ([`init(truncatingIfNeeded:)`](https://developer.apple.com/documentation/swift/binaryinteger/2925529-init)):
+  Trunca el ancho del tipo destino 
+
+La estrategia correcta la dicta el contexto, en algunas ocasiones es preferible prensar los valores a un rango que pueda ser representado, en otras es mejor no tener un valor del todo. En el caso de `numericCast(_:)`, el chequeo de rango se usa por conveniencia. La desventaja de este enfoque es que al invocar esta función con valores fuera de rango produce un error en tiempo de ejecución, específicamente, produce un `trap` en `overflow` para `-O` y `-Onone`.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: ejecución. Específicamente, produce...

File: 2018-10-15-numericcast.md [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+Este mismo tema se cubre a fondo en [Flight School Guide to Numbers](https://gumroad.com/l/swift-numbers).
+
+{% endinfo %}
+
+## Pensando literal y críticamente
+
+Antes de seguir, tomemos un momento para discutir enteros literales.
+
+Hemos [discutido en artículos anteriores](https://nshipster.com/swift-literals/) sobre la manera conveniente y extensible que nos brinda Swift para representar valores en nuestro código.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: sobre las variadas formas que nos brinda Swift para representar valores en nuestro código?

File: 2018-10-15-numericcast.md [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+Esto, combinado con la inferencia de tipos provista por el lenguaje, la mayoría del tiempo, las cosas "simplemente funcionan", lo que es maravilloso, excepto en las confusas ocasiones en que "simplemente no funcionan".
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: what about using spanish quotes? « »

File: 2018-10-15-numericcast.md [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+Usemos el siguiente ejemplo de arrays, con y sin signo, que contienen idénticos valores literales:
+
+```swift
+let arrayOfInt: [Int] = [1, 2, 3]
+let arrayOfUInt: [UInt] = [1, 2, 3]
+```
+
+Apesar de su equivalente apariencia, no podemos hacer esto:
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: A pesar

File: 2018-10-15-numericcast.md [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+```swift
+arrayOfInt as [UInt] // Error: Cannot convert value of type '[Int]' to type '[UInt]' in coercion
+```
+
+Una forma de sobrepasar este problema sería pasar la función `numericCast` como parámetro a `map(_:)`:
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: solventar

File: 2018-10-15-numericcast.md [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+Asemejando una funcionalidad en tiempo de ejecución a otra en tiempo de compilación, `numericCast (_ :)` está más cerca de `as!` que `as` o` as? `.
+
+Comparemos este comportamiento a lo que sucede si usamos el inicializador de conversión exacta `init?(exactly:)`:
+
+```swift
+let arrayOfNegativeInt: [Int] = [-1, -2, -3]
+arrayOfNegativeInt.map(UInt.init(exactly:)) // [nil, nil, nil]
+```
+
+`numericCast(_:)` al igual que su conversión con chequeo de rango es una herramienta contundente, por lo que es importante entender las concesiones que se hacen al usarlo.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: `numericCast(_:)`, al igual que su conversión con chequeo de rango, es una herramienta contundente, por lo que es importante entender las concesiones que se hacen al usarlo.

File: 2018-10-15-numericcast.md [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+## El costo de tener la razón
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: El coste?

File: 2018-10-15-numericcast.md [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+
+Generalmente en Swift se recomienda usar `Int` para enteros (y `Double` para valores de punto flotante) a menos que exista una *muy* buena razón para hacerlo.
+
+La propiedad `count` de una ``Collection`` nunca será, por definición, un valor negativo, sin embargo se usa un `Int` en vez de un `UInt` por que el costo de los viajes ida y vuelta entre tipos al interactuar con otras `API`s es más alto que el uso de un tipo más preciso.
+
+Por la misma razón, casi siempre es mejor usar `Int` para representar números pequeños, como [días de la semana](https://nshipster.com/datecomponents/), a pesar que cualquier otro valor estaría contenido, con espacio de sobra, en un tipo de 8 bits.
+
+El argumento más sólido para esta práctica es usar Swift para tener una discusión de 5 minutos con una `API` en C.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: No se entiende muy bien esta frase

File: 2018-10-15-numericcast.md [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+
+```swift
+import Darwin
+
+func random(in range: Range<Int>) -> Int {
+    return numericCast(arc4random_uniform(numericCast(range.count))) + range.lowerBound
+}
+```
+
+Podemos ver que `numericCast(_:)` no está realizando ninguna tarea que no fuese posible mediante el correcto inicializadores para cada tipo, al contrario, está siendo usado como un indicador de la naturaleza superflua de la conversión, la cantidad mínima de código para compilar.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: No se entiende bien este párrafo

File: 2018-10-15-numericcast.md [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+
+```swift
+random(in: 0..<0x1_0000_0000) // 🧞‍ Fatal error: Not enough bits to represent the passed value
+```
+
+Si observamos la [implementación de la librería estándar](https://github.com/apple/swift/blob/7f7b4f12d3138c5c259547c49c3b41415cd4206e/stdlib/public/core/Integers.swift#L2537-L2560), la que nos permite usar `Int.random(in: 0...10)`, veremos que usa una prensa, en vez una conversión de chequeo de rango. También vemos que en vez de delegar su trabajo a un método como `arc4random_uniform`, [llena una serie de valores desde un buffer hacia bytes aleatorios](https://github.com/apple/swift/blob/7f7b4f12d3138c5c259547c49c3b41415cd4206e/stdlib/public/core/Random.swift#L156-L177).
+
+---
+
+Lograr que el código compile es distinto a hacer las cosas bien, pero a veces ocupamos lo primero para llegar a lo segundo. Cuando se usa cuidadosamente `numericCast(_:)` es una util herramienta para resolver problemas rápidamente.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: herramienta útil para resolver problemas con rapidez.

File: 2018-10-15-numericcast.md [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+
+Un segundo beneficio es el de indicarnos posibles comportamientos erróneos más claramente que al usar inicializador convencionales.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: inicializadores

File: 2018-10-15-numericcast.md [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+
+Al final del día, la programación se trata en describir _exactamente_ lo que queremos, usualmente con un nivel obsesivo de detalle.
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: Al final, la programación no es otra cosa que describir de manera _exacta_ lo que queremos. Normalmente con un nivel de detalle obsesivo.

File: 2018-10-15-numericcast.md [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+
+No hay una instrucción para el CPU que sea equivalente para "haz lo correcto" (y ¿aunque lo existiera, [la usaríamos](https://github.com/FixIssue/FixCode)?). Afortunadamente Swift nos permite hacerlo de una manera segura y más concisa que en muchos otros lenguajes y honestamente ¿qué más podríamos desear?
```
Comment (jfsagasti) [inline 2018-10-15-numericcast.md]: y, honestamente, ¿

**Comment by jfsagasti** [comment]
Hi Leo! Happy New Year to you too :). Thanks for taking your time with another article. I've left my comments 👍

File: 2018-10-15-numericcast.md (hunk 1 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: You are right, necesitar is better, changing, thanks.

File: 2018-10-15-numericcast.md (hunk 2 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Better with similar, thanks.

File: 2018-10-15-numericcast.md (hunk 3 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Brinden makes more sense, good catch, thanks.

Updated the second bit too, inevitablemente was too wordy.

File: 2018-10-15-numericcast.md (hunk 4 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Fixed, thanks

File: 2018-10-15-numericcast.md (hunk 5 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Updated, thanks

File: 2018-10-15-numericcast.md (hunk 7 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Do you feel strongly about this one? I think it's mostly preferential.

File: 2018-10-15-numericcast.md (hunk 8 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: How about "sin importar talla ó tamaño"?

File: 2018-10-15-numericcast.md (hunk 9 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Fixed, thanks.

How do you feel about using `trap`? I couldn't find a suitable translation.

File: 2018-10-15-numericcast.md (hunk 10 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: I think some of the original meaning gets lost, how about this:
"...sobre las distintas maneras, tanto convenientes como extensibles, que nos..."?

File: 2018-10-15-numericcast.md (hunk 11 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: I think that might be too local? News sites like CNN en Español and BBC Mundo all use English quotes and I think that's probably the norm in this side of the world.

File: 2018-10-15-numericcast.md (hunk 12 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Fixed, thanks!

File: 2018-10-15-numericcast.md (hunk 13 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Thank you, I couldn't come up with a good synonym.

File: 2018-10-15-numericcast.md (hunk 14 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Thanks for the commas, much needed. I ended up changing "usarlo" for "usarla" since I was talking about an "herramienta" here.

File: 2018-10-15-numericcast.md (hunk 15 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: I think costo is alright, do you feel strongly about this?

File: 2018-10-15-numericcast.md (hunk 16 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: I updated this to: "El caso de uso perfecto para justificar esta práctica es usar Swift para interactuar con una `API` en C.", what do you think?

File: 2018-10-15-numericcast.md (hunk 17 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Updated to:
"Podemos ver que `numericCast(_:)` no está realizando ninguna tarea que no fuese posible mediante el correcto inicializadores de cada tipo, al contrario, está siendo usado como un indicador de la conversión, es efectivamente, la cantidad mínima de código para compilar."
What do you think?

File: 2018-10-15-numericcast.md (hunk 18 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Fixed, thanks.

File: 2018-10-15-numericcast.md (hunk 19 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: Fixed, thanks.

File: 2018-10-15-numericcast.md (hunk 20 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: How about meeting me half-way?
"Al fin y al cabo, la programación no es otra cosa que describir de manera exacta lo que queremos. Normalmente con un nivel de detalle obsesivo."

File: 2018-10-15-numericcast.md (hunk 21 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: I'll defer to you with it comes down to signos de puntuación :D, thanks

**Comment by leopic** [comment]
Sorry for the delay here, PR has been updated and I left comments for the items that were not updated, thanks @jfsagasti

File: 2018-10-15-numericcast.md (hunk 11 above)
Comment (jfsagasti) [reply 2018-10-15-numericcast.md]: Good to know, thanks! :)

File: 2018-10-15-numericcast.md (hunk 17 above)
Comment (jfsagasti) [reply 2018-10-15-numericcast.md]: el correcto inicializador de cada tipo. Al contrario, está...

I think it feels better that way. Let me know what you think :)

**Comment by jfsagasti** [comment]
Thanks Leo! I've left another little suggestion :)

File: 2018-10-15-numericcast.md (hunk 6 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: You are right about tuteo, forgot about that, I'll put it in the wiki, thank you Juan

File: 2018-10-15-numericcast.md (hunk 17 above)
Comment (leopic) [reply 2018-10-15-numericcast.md]: It does read way better, thank you!

**Comment by leopic** [comment]
Juan, thank you for your suggestions, I've implemented them and the article is now ready to be reviewed again.

**Comment by jfsagasti** [comment]
LGTM! Thank you Leo :)

**Comment by leopic** [comment]
All yours @mattt

**Comment by mattt** [comment]
Awesome! Thanks for your great work on this @leopic. And thank you @jfsagasti for taking the time to review everything.

I'll let @jfsagasti do the honors of merging this in and kickstarting auto-publication.
